arch/arm/core/cortex_m/mpu/arm_core_mpu.c

```diff
@@ -10,105 +10,294 @@
+	 * initialization.
+	 */
+	const struct k_mem_partition dyn_region_areas[] = {
+#if defined(CONFIG_APPLICATION_MEMORY)
+	/* Dynamic areas are also allowed in Application Memory. */
+		{
+		.start = (u32_t)&__app_ram_start,
+		.size = (u32_t)&__app_ram_end - (u32_t)&__app_ram_start,
+		},
+#endif /* CONFIG_APPLICATION_MEMORY */
+		{
+		.start = _MPU_DYNAMIC_REGIONS_AREA_START,
+		.size =  _MPU_DYNAMIC_REGIONS_AREA_SIZE,
+		}
+	};
+	arm_core_mpu_mark_areas_for_dynamic_regions(dyn_region_areas,
+		ARRAY_SIZE(dyn_region_areas));
+#endif /* CONFIG_MPU_REQUIRES_NON_OVERLAPPING_REGIONS */
 }
 
-/*
- * @brief Configure MPU memory domain
+/**
+ * @brief Use the HW-specific MPU driver to program
+ *        the dynamic MPU regions.
  *
- * This function configures per thread memory domain reprogramming the MPU.
- * The functionality is meant to be used during context switch.
+ * Program the dynamic MPU regions using the HW-specific MPU
+ * driver. This function is meant to be invoked every time the
+ * memory map is to be re-programmed, e.g during thread context
+ * switch, entering user mode, reconfiguring memory domain, etc.
  *
- * @param thread thread info data structure.
+ * For some MPU architectures, such as the unmodified ARMv8-M MPU,
+ * the function must execute with MPU enabled.
  */
-void configure_mpu_mem_domain(struct k_thread *thread)
+void _arch_configure_dynamic_mpu_regions(struct k_thread *thread)
 {
+	/* Define an array of k_mem_partition objects to hold the configuration
+	 * of the respective dynamic MPU regions to be programmed for
+	 * the given thread. The array of partitions (along with its
+	 * actual size) will be supplied to the underlying MPU driver.
+	 */
+	struct k_mem_partition dynamic_regions[_MAX_DYNAMIC_MPU_REGIONS_NUM];
+
+	u8_t region_num = 0;
+
+#if defined(CONFIG_USERSPACE)
+	/* Memory domain */
 	LOG_DBG("configure thread %p's domain", thread);
-	arm_core_mpu_disable();
-	arm_core_mpu_configure_mem_domain(thread->mem_domain_info.mem_domain);
-	arm_core_mpu_enable();
-}
+	struct k_mem_domain *mem_domain = thread->mem_domain_info.mem_domain;
+	if (mem_domain) {
+		LOG_DBG("configure domain: %p", mem_domain);
+		u32_t num_partitions = mem_domain->num_partitions;
+		struct k_mem_partition partition;
+		int i;
 
-void _arch_mem_domain_configure(struct k_thread *thread)
-{
-	configure_mpu_mem_domain(thread);
+		LOG_DBG("configure domain: %p", mem_domain);
+
+		for (i = 0; i < CONFIG_MAX_DOMAIN_PARTITIONS; i++) {
+			partition = mem_domain->partitions[i];
+			if (partition.size == 0) {
+				/* Zero size indicates a non-existing
+				 * memory partition.
+				 */
+				continue;
+			}
+			LOG_DBG("set region 0x%x 0x%x",
+				partition.start, partition.size);
+			dynamic_regions[region_num].start = partition.start;
+			dynamic_regions[region_num].size = partition.size;
+			dynamic_regions[region_num].attr = partition.attr;
+
+			region_num++;
+			num_partitions--;
+			if (num_partitions == 0) {
+				break;
```

Comment:
So:
- `i `iterates on the input array of partition items (some elements of this array may be null entries)
- `region_num `increments and populates the local array of dynamic memory map. We can't really use i for this, because of the possible null entries on the input array.

>early exit above (that maybe could be eliminated?)

Actually, I would like to keep it, so I do not run the FOR loop until `CONFIG_MAX_DOMAIN_PARTITIONS`, in case all non-null elements are already exhausted.

The num_partitions (the variable and this check) is the price we have to pay, to be able to terminate early the loop (this makes some sense for MPU hardware with quite many HW regions, e.g. 32 but only a few actually configured partitions)

Reply:
(addressed)